src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs

```diff
@@ -0,0 +1,329 @@
+
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Force the matrix factorization P and Q to be non-negative", ShortName = "nn")]
+            public bool NonNegative;
+        };
+
+        internal const string Summary = "From pairs of row/column indices and a value of a matrix, this trains a predictor capable of filling in unknown entries of the matrix, "
+            + "utilizing a low-rank matrix factorization. This technique is often used in recommender system, where the row and column indices indicate users and items, "
+            + "and the value of the matrix is some rating. ";
+
+        private readonly Double _lambda;
+        private readonly int _k;
+        private readonly int _iter;
+        private readonly Double _eta;
+        private readonly int _threads;
+        private readonly bool _quiet;
+        private readonly bool _doNmf;
+
+        public override PredictionKind PredictionKind => PredictionKind.Recommendation;
+        public const string LoadNameValue = "MatrixFactorization";
+
+        /// <summary>
+        /// The row, column, and label columns that the trainer expects. This module uses tuples of (row index, column index, label value) to specify a matrix.
+        /// For example, a 2-by-2 matrix
+        ///   [9, 4]
+        ///   [8, 7]
+        /// can be encoded as tuples (0, 0, 9), (0, 1, 4), (1, 0, 8), and (1, 1, 7). It means that the row/column/label column contains [0, 0, 1, 1]/
+        /// [0, 1, 0, 1]/[9, 4, 8, 7]. Note that for a given matrix, row indices are column indices are denoted by Y and X, respectively.
+        /// </summary>
+        public readonly SchemaShape.Column XColumn; // column indices of the training matrix
+        public readonly SchemaShape.Column YColumn; // row indices of the training matrix
+        public readonly SchemaShape.Column LabelColumn;
+
+        /// <summary>
+        /// The <see cref="TrainerInfo"/> contains general parameters for this trainer.
+        /// </summary>
+        public override TrainerInfo Info { get; }
+
+        /// <summary>
+        /// Extra information the trainer can use. For example, its validation set (if not null) can be use to evaluate the
+        /// training progress made at each training iteration.
+        /// </summary>
+        public readonly TrainerEstimatorContext Context;
+
+        /// <summary>
+        /// Legacy constructor initializing a new instance of <see cref="MatrixFactorizationTrainer"/> through the legacy
+        /// <see cref="Arguments"/> class.
+        /// </summary>
+        /// <param name="env">The private instance of <see cref="IHostEnvironment"/>.</param>
+        /// <param name="args">An instance of the legacy <see cref="Arguments"/> to apply advanced parameters to the algorithm.</param>
+        public MatrixFactorizationTrainer(IHostEnvironment env, Arguments args) : base(env, LoadNameValue)
+        {
+
+            const string posError = "Parameter must be positive";
```

Comment:
you don't need this line.

Reply:
It can save several characters because several places are using the same message.

Reply:
I meant empty line above, sorry, misselect.

---
In reply to: [227614661](https://github.com/dotnet/machinelearning/pull/1263#discussion_r227614661) [](ancestors = 227614661)

Reply:
I see. Fixed. Thanks.

---
In reply to: [227891062](https://github.com/dotnet/machinelearning/pull/1263#discussion_r227891062) [](ancestors = 227891062,227614661)